src/Chainweb/Pact/TransactionExec.hs

```diff
@@ -294,117 +306,108 @@ applyLocal
-      applyExec env interp (cmdToRequestKey c) pm (_pSigners _cmdPayload) (toUntypedHash _cmdHash) txLogs nsp
+      applyExec interp rk pm signers chash nsp
     Continuation ym ->
-      applyContinuation env interp (cmdToRequestKey c) ym (_pSigners _cmdPayload) (toUntypedHash _cmdHash) txLogs nsp
+      applyContinuation interp rk ym signers chash nsp
+  where
+    rk = cmdToRequestKey cmd
+    signers = _pSigners $ _cmdPayload cmd
+    chash = toUntypedHash $ _cmdHash cmd
+    payload = _pPayload $ _cmdPayload cmd
 
 -- | Execute an 'ExecMsg' and Return the result with module cache
 --
 applyExec
-    :: CommandEnv p
-    -> Interpreter p
+    :: Interpreter p
     -> RequestKey
     -> ExecMsg ParsedCode
     -> [Signer]
     -> Hash
-    -> [TxLog Value]
     -> NamespacePolicy
-    -> IO (T2 (CommandResult [TxLog Value]) ModuleCache)
-applyExec env@CommandEnv{..} interp rk em senderSigs hsh prevLogs nsp = do
-    EvalResult{..} <- applyExec' env interp em senderSigs hsh nsp
+    -> TransactionM p (CommandResult [TxLog Value])
+applyExec interp rk em senderSigs hsh nsp = do
+    EvalResult{..} <- applyExec' interp em senderSigs hsh nsp
+    logs <- use transactionLogs
     -- applyExec enforces non-empty expression set so `last` ok
-    return $! T2 (CommandResult rk _erTxId (PactResult (Right (last _erOutput)))
-      _erGas (Just $ prevLogs <> _erLogs) _erExec Nothing) _erLoadedModules -- TODO add perf metadata
+    return $! CommandResult rk _erTxId (PactResult (Right (last _erOutput)))
+      _erGas (Just logs) _erExec Nothing
 
 -- | Variation on 'applyExec' that returns 'EvalResult' as opposed to
 -- wrapping it up in a JSON result.
 --
 applyExec'
-    :: CommandEnv p
-    -> Interpreter p
+    :: Interpreter p
     -> ExecMsg ParsedCode
     -> [Signer]
     -> Hash
     -> NamespacePolicy
-    -> IO EvalResult
-applyExec' CommandEnv{..} interp (ExecMsg parsedCode execData) senderSigs hsh nsp = do
-    -- fail if parsed code contains no expressions
-    --
-    when (null $ _pcExps parsedCode) $
-      throwCmdEx "No expressions found"
+    -> TransactionM p EvalResult
+applyExec' interp (ExecMsg parsedCode execData) senderSigs hsh nsp
+    | null (_pcExps parsedCode) = throwCmdEx "No expressions found"
+    | otherwise = do
 
-    er <- evalExec senderSigs interp evalEnv parsedCode
+      tenv <- ask
+      genv <- use transactionGasEnv
 
-    for_ (_erExec er) $ \PactExec{..} -> logLog _ceLogger "DEBUG"
-      $ "applyExec: new pact added: "
-      <> show (_pePactId, _peStep, _peYield, _peExecuted)
+      let eenv = evalEnv tenv genv
+      er <- liftIO $! evalExec senderSigs interp eenv parsedCode
 
-    return er
+      liftIO $! for_ (_erExec er) $ \pe -> logLog (_txLogger tenv) "DEBUG"
+        $ "applyExec: new pact added: "
+        <> show (_pePactId pe, _peStep pe, _peYield pe, _peExecuted pe)
+
+      -- set log + cache updates
+      transactionLogs <>= (_erLogs er)
+      transactionCache .= (_erLoadedModules er)
```

Comment:
Review code that is manually adding stuff to this, e.g. in `runPayload` as it seems to happen automatically here